### ISSUE TITLE
std::lock_guard on member variable during destruction 

### ISSUE THREAD

**Issue by danginsburg** [KhronosGroup/MoltenVK#174]
I've been looking into some difficult to repro crashes in Dota 2:

https://github.com/ValveSoftware/Dota-2-Vulkan/issues/249
https://github.com/ValveSoftware/Dota-2-Vulkan/issues/246

From studying minidumps, a lot of the crashes happen when destroying a VkFence or VkSemaphore.  I discovered that if I run with Xcode Diagnostics "Malloc Scribble" and "Malloc Guard Edges" I get an assertion in MVKMetalCompiler::destroy() which does this:

```
void MVKMetalCompiler::destroy() {
	lock_guard<mutex> lock(_completionLock);

	_isDestroyed = true;
	maybeDestroy();
}
```
The problem is that the _completionLock is a member variable and is destroyed in the case maybeDestroy() deletes "this".  I notice that MVKFence and MVKSemaphore have similar patterns (and other places in the code).

There needs to be a different approach to mutex locking during destruction if it is needed.

**Comment by billhollings** [comment]
@danginsburg 

This synchronization issue appeared elsewhere and was fixed in PR #165 a couple of weeks ago.

**Comment by danginsburg** [comment]
Thanks, reports are indicating that the updated MoltenVK is fixing the crashes.  Closing.
